src/lib/AuthManager.js

```diff
@@ -0,0 +1,236 @@
+import browser from './browser-api'
+
+// An AuthSession can be used to make API fetch() calls to Nextcloud. It 
+// overrides the requests headers to reduce the number of times, the server
+// has to validate passwords.
+// In order to do that without interfering with the users session cookies, 
+// it intercepts relevant requests/responses and stores the cookies in a dedicated 
+// CookieJar.
+//
+// Because a CookieJar does not respect any attributes given to 
+// cookies (like "secure" or "expires") and because session timeouts are not
+// detected, the follwing rules apply:
+//  - Keep the lifetime of an AuthSession short!
+//  - Never save an AuthSession to persistent or otherwise insecure memory!
+export class AuthSession {
+
+  // @param "https://example.org" This session can only be used for API calls
+  //          to this server.
+  constructor(serverUrl) {
+    if (serverUrl === undefined) {
+      throw "AuthSession constructor arg undefined: serverUrl"
+    }
+    this.authman = window.authManager
+    if (this.authman === undefined) {
+      throw "No AuthManager available."
+    }
+    this.cookieJar = new CookieJar()
+    this.id = this.authman.newSessionId()
+    this.serverUrl = serverUrl
+    this.ownedRequests = {} // dict: request.id => true
+    this.authman.addSession(this)
+
+    // it looks like we only need to send auth 2 times for nextcloud? (we send 3 times now)
+    this.needsAuthentication = true 
+    // some browsers might not allow us to acces relevant httpOnly cookies
+    this.everRecievedCookie = false
+  }
+
+  destructor() {
+    this.authman.removeSession(this)
+    let id = this.id
+    Object.keys(this).forEach(key => delete this[key])
```

Comment:
this shouldn't be necessary, I think. Garbage collection ftw :)

Reply:
This way when someone tries to use a destructed AuthSession, he will get an error for `authSession.* is undefined` - then he will look at the authSession object and will see that `authSession.info = "is destroyed"`. Otherwise you can still accidentally use destructed Sessions and get weird unrelated errors. 🤷

Reply:
I think i should add a comment explaining stupid stuff like that.

Reply:
Ah, I see. We could make this explicit, by just setting a flag in the destructor and check when the session is used. That way we even get an explicit error message :)